model/vcs/metrics.yaml

```
@@ -0,0 +1,151 @@
+      - ref: vcs.repository.url.full
+        requirement_level: required
+  - id: metric.vcs.change.duration
+    type: metric
+    metric_name: vcs.change.duration
+    brief: 'The time duration a change (pull request/merge request/changelist) has been in a given state.'
+    instrument: gauge
+    unit: "s"
+    stability: experimental
+    attributes:
+      - ref: vcs.repository.url.full
+        requirement_level: required
+      - ref: vcs.ref.name
+        requirement_level: required
+      - ref: vcs.change.state
+        requirement_level: required
+  - id: metric.vcs.change.time_to_approval
+    type: metric
+    metric_name: vcs.change.time_to_approval
+    brief: 'The amount of time it took a change (pull request/merge request/changelist) to get the first approval'
+    instrument: gauge
+    unit: "s"
+    stability: experimental
+    attributes:
+      - ref: vcs.repository.url.full
+        requirement_level: required
+      - ref: vcs.ref.name
+        requirement_level: required
+  - id: metric.vcs.repository.count
+    type: metric
+    metric_name: vcs.repository.count
+    brief: 'The number of repositories in an organization'
+    instrument: updowncounter
+    unit: "{repository}"
+    stability: experimental
+    attributes: []
+  - id: metric.vcs.ref.count
+    type: metric
+    metric_name: vcs.ref.count
+    brief: 'The number of refs of type branch or tag in a repository'
+    instrument: updowncounter
+    unit: "{ref}"
+    stability: experimental
+    attributes:
+      - ref: vcs.repository.url.full
+        requirement_level: required
+      - ref: vcs.ref.type
+        requirement_level: required
+  - id: metric.vcs.change.lines_delta
+    type: metric
+    metric_name: vcs.change.lines_delta
+    brief: 'The number of lines added/removed in a change relative to the common ancestor with the target branch.'
+    note: 'This metric should be reported for each `vcs.line_change.type` value. For example if a change added 3 lines and removed 2 lines, instrumentation SHOULD report two measurements: 3 and 2 (both positive numbers).'
+    instrument: gauge
+    unit: "{line}"
+    stability: experimental
+    attributes:
+      - ref: vcs.repository.url.full
+        requirement_level: required
+      - ref: vcs.ref.name
+        requirement_level: required
+      - ref: vcs.ref.type
+        requirement_level: required
+      - ref: vcs.line_change.type
+        requirement_level: required
+  - id: metric.vcs.ref.lines_delta
+    type: metric
+    metric_name: vcs.ref.lines_delta
+    brief: 'The number of lines added/removed in a ref (branch) relative to the previous day'
+    note: 'This metric should be reported for each `vcs.line_change.type` value. For example if a ref added 3 lines and removed 2 lines, instrumentation SHOULD report two measurements: 3 and 2 (both positive numbers).'
+    instrument: gauge
+    unit: "{line}"
+    stability: experimental
+    attributes:
+      - ref: vcs.repository.url.full
```

I would add these two attributes to the `vcs.ref.lines_delta` metric.
With this can optionally add the vcs.change.id if we are reporting the metric for a change.
We explicitly state against which other branch we compare the revisions against (new attribute).
This could the default branch for other branches and the target branch for a change.
```suggestion
    attributes:
      - ref: vcs.change.id
        requirement_level:
          conditionally_required: if a change is associate with the ref.
      - ref: vcs.ref.compare_to
        requirement_level: required
      - ref: vcs.repository.url.full
```
How would we specify `vcs.ref.compare_to` so that for the default branch we compare against the start of time?

---

I'm glad you took a stab at adding an attribute. I was thinking about that this weekend. What we're missing here is context! (attributes). For `ref` to `ref` comparison, we have to compare the `HEAD.*` with the `BASE.*`. In events I've seen, usually `BASE` and `HEAD` are separate namespaces that each have `ref` and `sha` within them. Mercurial & SVN might have slight nuances. 

I'm currently thinking that we should likely include attributes like the following. 

* `vcs.ref.head.name`
* `vcs.ref.head.sha`
* `vcs.ref.base.name`
* `vcs.ref.base.sha`

or we can adjust the current attributes from `vcs.ref.*` to be

* `vcs.base.ref.name`
* `vcs.base.ref.revision`
* `vcs.base.ref.type`
* `vcs.head.ref.name`
* `vcs.head.ref.revision`
* `vcs.head.ref.type`

Thoughts?

---

I would rename the following attributes
* `vcs.ref.name`
* `vcs.ref.revision`
* `vcs.ref.type`
to
* `vcs.ref.head.name`
* `vcs.ref.head.revision`
* `vcs.ref.head.type`
including for all the metrics.
For the `vcs.ref.count` metric I am not sure whether to rename the `vcs.ref.type` attribute.
Should I rename to be consistent with the other ref attributes or keep it as `vcs.ref.type`?

I would add
* `vcs.ref.base.name`
* `vcs.ref.base.revision`
* `vcs.ref.base.type`
for use in the `revisions_delta` and `lines_delta` metrics.

---

Given the proposed change in https://github.com/open-telemetry/semantic-conventions/pull/1521 https://github.com/open-telemetry/semantic-conventions/blob/3207d0f250b03caed33c35a1980d3a3d9abc2d5b/docs/general/events.md?plain=1#L101-L104
we should find an attribute name other than `vcs.ref.*`.

---

Feel free to suggest on https://github.com/open-telemetry/semantic-conventions/pull/1521 a name other than `ref` for these external references. Maybe we can come up with an alternative name that won't clash.

---

I have implemented the new attributes and combined `vcs.ref.lines_delta` and `vcs.change.lines_delta` in 8f8bb73943ce328f172f402b0683b45412073c2f.